bowlrelease/runner/trainer.py

```diff
@@ -24,73 +26,128 @@ def train(dataloader, model, loss_fn, optimizer, device):
+        with open(
+            os.path.join(log_path, "test_predictions.json"),
+            "w",
+            encoding="utf-8",
+        ) as f_json:
+            json.dump(pr_events, f_json, ensure_ascii=False, indent=4)
+        with open(
+            os.path.join(log_path, "test_groundtruths.json"),
+            "w",
+            encoding="utf-8",
+        ) as f_json:
+            json.dump(gt_events, f_json, ensure_ascii=False, indent=4)
+
+    return pq_test
+
+
+def inference(dataloader, model, device, log_path=""):
+    """Inference function"""
+    size = len(dataloader.dataset)
+    size *= dataloader.dataset.length_seq
+    model.eval()
+
+    pred_dict = {}
+
+    with torch.no_grad():
+        for idx, X in enumerate(dataloader):
+            video_batch_list = [
+                dataloader.dataset.indxs[i][0]
+                for i in range(len(X) * idx, len(X) * (idx + 1))
+            ]
+            X = X.to(device)
+            pred = model(X)
+
+            pred_np = (pred > 0.5).type(torch.float).cpu().numpy()
+
+            for ib, vid in enumerate(video_batch_list):
+                pred_dict.setdefault(vid, []).append(pred_np[ib, :])
+
+    for k, v in pred_dict.items():
+        pred_dict[k] = np.concatenate(v)
+    pr_events = {}
+    for k, val in pred_dict.items():
+        events_ = rising_edge(val)
+        pr_events[k] = dict(enumerate(events_))
+    LOGGER.info("Inference run on the Challenge set.\n")
+    if log_path:
+        with open(
+            os.path.join(log_path, "challenge_predictions.json"),
+            "w",
+            encoding="utf-8",
+        ) as f_json:
+            json.dump(pr_events, f_json, ensure_ascii=False, indent=4)
+        LOGGER.info(
+            f"Predictions saved in {log_path}/challenge_predictions.json.\n"
+        )
 
 
 def compute_metric(preds, gts):
     """Compute the metric and apply weights to final frame"""
     weights = np.zeros_like(gts)
-    for id, gt in enumerate(gts):
-        if gt == 1:
+    for id_, gt_ in enumerate(gts):
+        if gt_ == 1:
             weights[id] = 1
-            if id < gts.size and gts[id + 1] == 0:
+            if id_ < gts.size and gts[id_ + 1] == 0:
                 weights[id] = 2
     avpr = iou_metric(preds, gts, weights, 0.5)
     LOGGER.info(f"IoU Metric: \n {(100*avpr):>0.1f}% \n")
     return avpr
 
 
-def get_loss_and_optimizer(model, features):
+def get_loss_and_optimizer(model):
     """Return loss and optimizer"""
-    if features:
-        loss_fn = torch.nn.MSELoss()
-        optimizer = torch.optim.AdamW(model.parameters(), amsgrad=True)
-    else:
-        loss_fn = torch.nn.CrossEntropyLoss()
-        optimizer = torch.optim.AdamW(
-            [
-                {"params": model.rn_model.parameters(), "lr": 1e-5},
-                {"params": model.feat_bn.parameters()},
-                {"params": model.classifier.parameters()},
-            ],
-            lr=1e-3,
-        )
+
+    loss_fn = torch.nn.MSELoss()
```

Comment:
Why MSE Loss if you are doing binary classification

Reply:
moved to bce